rs/bitcoin/adapter/src/lib.rs

```diff
@@ -151,33 +159,61 @@ pub struct AdapterState {
     /// This way the adapter would always be in idle when starting since 'elapsed()' is greater than 'idle_seconds'.
     /// On MacOS this approach caused issues since on MacOS Instant::now() is time since boot and when subtracting
     /// 'idle_seconds' we encountered an underflow and panicked.
-    last_received_at: Arc<RwLock<Option<Instant>>>,
-    /// The field contains how long the adapter should wait to before becoming idle.
-    idle_seconds: u64,
+    ///
+    /// I't simportant that this value is set to None on startup.
+    last_received_tx: watch::Sender<Option<Instant>>,
 }
 
 impl AdapterState {
-    /// Crates new instance of the AdapterState.
+    /// Creates a new instance of the AdapterState.
     pub fn new(idle_seconds: u64) -> Self {
+        // Initialize the watch channel with `None`, indicating no requests have been received yet.
+        let (last_received_tx, _) = watch::channel(None);
         Self {
-            last_received_at: Arc::new(RwLock::new(None)),
             idle_seconds,
+            last_received_tx,
         }
     }
 
-    /// Returns if the adapter is idle.
-    pub fn is_idle(&self) -> bool {
-        match *self.last_received_at.read() {
-            Some(last) => last.elapsed().as_secs() > self.idle_seconds,
-            // Nothing received yet still in idle from startup.
-            None => true,
+    /// Updates the state to indicate a request was received now.
+    pub fn received_now(&self) {
+        let _ = self.last_received_tx.send(Some(Instant::now()));
+    }
+
+    /// A future that returns when/if the adapter becomes/is awake.
+    pub async fn become_awake(&self) {
+        let mut last_received_rx = self.last_received_tx.subscribe();
+        loop {
+            if let Some(last) = *last_received_rx.borrow() {
+                if last.elapsed().as_secs() < self.idle_seconds {
+                    return;
+                }
+            }
+            // Wait for a change in the last received time.
+            let _ = last_received_rx.changed().await;
         }
```

Comment:
You can use https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.wait_for here, and write something like:
```suggestion
        last_received_rx.wait_for(|last| {
          last.is_some_and(|last| {  
             last.elapsed().as_secs()>= self.idle_seconds
          })
        }).await;
```

Reply:
thanks